File: python-code/broadcast_reduce.py
from mpi4py import MPI
import numpy as np

comm = MPI.COMM_WORLD
rank = comm.rank
size = comm.size

if rank == 0:
    data = np.arange(1, size + 1)
else:
    data = None

# Scatter the data array to all processes
data = comm.bcast(data, root=0)
print("Process", rank, "has received ", data)

# Reduce the local sum to obtain the global sum
global_sum = comm.reduce(data, op=MPI.SUM, root=0)

# Print the result on the root process
if rank == 0:
    print("The sum of the array is", global_sum)

File: c-code/broadcast_reduce.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char** argv) {
    int rank, size;
    int root = 0;
    int data = 0;
    int sum = 0;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // Broadcast data from root process to all other processes
    MPI_Bcast(&data, 1, MPI_INT, root, MPI_COMM_WORLD);
    printf("Process %d received broadcast data: %d\n", rank, data);

    // Perform reduction operation on data from all processes
    MPI_Reduce(&data, &sum, 1, MPI_INT, MPI_SUM, root, MPI_COMM_WORLD);
    if (rank == root) {
        printf("Sum of all data: %d\n", sum);
    }

    MPI_Finalize();
    return 0;
}
